src/layer.rs

```diff
@@ -30,6 +30,16 @@ pub trait Layer: Send + Debug + erased_serde::Serialize {
         bytes: &[u8],
     ) -> Result<(Option<Box<dyn Layer + Send>>, usize), Error>;
 
+    /// Main 'encoder' function.
+    ///
+    /// The return value is a `Vec<u8>` on success. This indicates the encoded packet of the 
+    /// layer. Typically return a [SculptingError][`crate::errors::Error::SculptingError`], 
+    /// but may as well return other values. This updates the fields of the packet using the
+    /// encoded bytes of the next layer(`next_layer`), and `info`, to provide additional details
+    /// about the packet. This can be the result of [Layer::name()][`Self::name()`]
+    /// of the next packet. 
+    fn stack_and_encode(&mut self, next_layer: &[u8], info: &str) -> Result<Vec<u8>, Error>;
```

Comment:
Wondering whether the `next_layer` above should be an `Option` (which will be in `None` in most cases, thus making this function a no-op`).

Reply:
I'd initially used an `Option<&Vec<u8>>`, but settled on this for more general cases. An non existing next layer can be represented as an empty slice. Or should I use `Option<&[u8]>` to make intent clearer?

Reply:
Yes, Option makes it more explicit, so a better idea. Yes, it should be `Option<&[u8]>`.